File: smart_campus/backend/internal/domain/entities/course.go
package entities

import (
	"time"

	"github.com/google/uuid"
)

// Course represents a course in the system
type Course struct {
	ID          string    `json:"id"`
	CourseCode  string    `json:"course_code"`
	CourseName  string    `json:"course_name"`
	Department  string    `json:"department"`
	YearOfStudy int       `json:"year_of_study"`
	Semester    int       `json:"semester"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// NewCourse creates a new course instance
func NewCourse(courseCode, courseName, department string, yearOfStudy, semester int) *Course {
	now := time.Now()
	return &Course{
		ID:          uuid.New().String(),
		CourseCode:  courseCode,
		CourseName:  courseName,
		Department:  department,
		YearOfStudy: yearOfStudy,
		Semester:    semester,
		CreatedAt:   now,
		UpdatedAt:   now,
	}
}

// Update updates the course details
func (c *Course) Update(courseName, department string, yearOfStudy, semester int) {
	c.CourseName = courseName
	c.Department = department
	c.YearOfStudy = yearOfStudy
	c.Semester = semester
	c.UpdatedAt = time.Now()
}

// IsValidForYear checks if the course is valid for a given year of study
func (c *Course) IsValidForYear(year int) bool {
	return c.YearOfStudy == year
}

// IsValidForDepartment checks if the course is valid for a given department
func (c *Course) IsValidForDepartment(department string) bool {
	return c.Department == department
}

File: smart_campus/lib/domain/entities/course.dart
import 'package:equatable/equatable.dart';

class Course extends Equatable {
  final String id;
  final String courseCode;
  final String courseName;
  final String department;
  final int yearOfStudy;
  final int semester;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Course({
    required this.id,
    required this.courseCode,
    required this.courseName,
    required this.department,
    required this.yearOfStudy,
    required this.semester,
    required this.createdAt,
    required this.updatedAt,
  });

  Course copyWith({
    String? id,
    String? courseCode,
    String? courseName,
    String? department,
    int? yearOfStudy,
    int? semester,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Course(
      id: id ?? this.id,
      courseCode: courseCode ?? this.courseCode,
      courseName: courseName ?? this.courseName,
      department: department ?? this.department,
      yearOfStudy: yearOfStudy ?? this.yearOfStudy,
      semester: semester ?? this.semester,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'courseCode': courseCode,
      'courseName': courseName,
      'department': department,
      'yearOfStudy': yearOfStudy,
      'semester': semester,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  factory Course.fromJson(Map<String, dynamic> json) {
    return Course(
      id: json['id'] as String,
      courseCode: json['courseCode'] as String,
      courseName: json['courseName'] as String,
      department: json['department'] as String,
      yearOfStudy: json['yearOfStudy'] as int,
      semester: json['semester'] as int,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  @override
  List<Object?> get props => [
        id,
        courseCode,
        courseName,
        department,
        yearOfStudy,
        semester,
        createdAt,
        updatedAt,
      ];
}
